src/journal/journald-stream.c

```diff
@@ -548,9 +548,20 @@ static int stdout_stream_process(sd_event_source *es, int fd, uint32_t revents,
         if (ucred && ucred->pid != s->ucred.pid) {
                 /* force out any previously half-written lines from a different process, before we switch to
                  * the new ucred structure for everything we just added */
-                r = stdout_stream_scan(s, true);
-                if (r < 0)
-                        goto terminate;
+                if (s->length > 0) {
+                        char saved_char;
+
+                        saved_char = s->buffer[s->length];
+                        s->buffer[s->length] = '\0';
+
+                        r = stdout_stream_line(s, s->buffer, LINE_BREAK_EOF);
+                        if (r < 0)
+                                goto terminate;
+
+                        s->buffer[s->length] = saved_char;
+                        memmove(s->buffer, &s->buffer[s->length], l);
+                        s->length = 0;
```

Comment:
This doesn't seem right. `stdout_stream_scan()` already has nearly identical code:
https://github.com/systemd/systemd/blob/master/src/journal/journald-stream.c#L473-L481
The difference is that the dance with `saved_char` is not done there. Instead of duplicating the logic here, please fix it there, so that the call to `stdout_stream_scan` here has the correct effect.

Reply:
This is why I split the function into 2 parts: flush and scan. In the case of flush we must save the char that is replaced by `\0` since there is at this position the character of the output of another PID.

Reply:
@keszybz I added a test case, could you review the change again after reading the following comment https://github.com/systemd/systemd/pull/15685#discussion_r420028460

I am almost sure that the overall change set is right, maybe not perfect and can be improved on the style, ...